### DISCUSSION TITLE
static_argnums and JAX control flows

### DISCUSSION THREAD

**Discussion by jecampagne** [jax-ml/jax#14818] [Q&A]
Hi, 
When one explicitly use jit as decorator of a function, he/she has the possibility to set some arguments as static.
But, when dealing with `jax.lax.scan(body,init,xs)`  (idem for other control flows methods). the compilation of the `body` function is done w/o any control. Is there a way to pass that certain element of `init`  are static?

**Comment by jakevdp** [comment]
There's no way to specify static variables here because scanned-over arguments cannot be static by definition (they change from iteration to iteration). The only way to have static variables in the scan body function is if they don't participate in the scan, and you can generally do this via closure, for example:
```python
import jax
from functools import partial

def f(c, x, flag):  # flag must be static
  if flag:
    return c + 1, x
  else:
    return c, x

jax.lax.scan(partial(f, flag=True), 0, None, 10)
```

**Reply by jecampagne** [reply]
Well, I have a tricky use-case with nested control flows but I need jacrev, so direct while_loop and also indirect through fori_loop are forbidden.

I will try to summarize although I have the code on Colab if necessary. The jec_fori_loop is just a C&P of the jax.lax.foi_loop to allow to print why scan is not used, from which I see that the lax.fori_loop uses the while_loop because of failure of this test
```python
  if (isinstance(core.get_aval(lower), ConcreteArray) and
      isinstance(core.get_aval(upper), ConcreteArray)):
```

So here is my nasty use-case...
 
```python

def func_inner(argt):
    #decode argt
    arr,... , (m,mwrap,step, nloops)= argt

    def step_once(val0):
        #decode val0
        i, ...., argt  = val0
        arr, .... , (m,mwrap,step, nloops) = argt
       
        i += 1
       
       def body_1(k,val1):
            #decode val1
            arr, i, ...,step = val1
            ....
           #encode val1
            return (arr, i,...,step)

       k = jax.lax.min(m-i, mwrap-2)
       val1_init = (arr, i, ...., step)
       #val1 = jax.lax.fori_loop(0,k,body_1, val1_init) # ok but certainly through while
       val1 = jec_fori_loop(0,k,body_1, val1_init)        # Nok scan failed due to Concrete
       arr, i, ... , step = val1

       ....
       #encode (end of step_once)
       argt = (arr, ... , (m,mwrap,step, nloops))
       val0 = (i, .... , argt)
       return val0

    # main part of fun_inner
    i = mwrap-1
    argt= (arr, ... , (m,mwrap,step, nloops) )
    
    def noop(carry):
         ### jax.debug.print("noop")
        return carry
    
   def body_fn(carry,dummy):
       (i, ... , argt) = carry
       (arr, .... , (m,mwrap,step, nloops) ) = argt
       return jax.lax.cond(i<m, step_once, noop, carry), None

   carry = (i, .... , argt)
   carry, _ = jax.lax.scan(body_fn, carry,nloops)

   i, ..., argt = carry
   arr, .... , (m,mwrap,step, nloops) = argt

  return arr, i

   
def func_outer(arr):

   def body_fn(carry,dummy):
        #decode carry
        (arr, ..., (m,mwrap,step, nloops), ...) = carry

       argt = arr, .... , (m,mwrap,step, nloops)
       arr, ideg = func_inner(argt)
       
       #encode carry
        carry = arr, (m,mwrap,step, nloops), ...
        return carry, None

  # define init carry variables
   carry = (arr, ..., (m,mwrap,step, nloops), ...)
   carry, _ = jax.lax.scan(body_fn, carry, None, length=1)
   arr  = carry[0]
   return arr
```

I think that what I need at the end of the day is the jacrev of func_outer wrt arr (2D array).

**Reply by jakevdp** [reply]
You'll need `nloops` to be static, and so you can't include it as part of the scan carry.  I would define your function something like this:
```python
def func_inner(argt, nloops=100):
    #decode argt
    arr,... , (m,mwrap,step)= argt
```
Since `nloops` doesn't change throughout your program, it should be fine to define it only once at the function level.

Will that work for you?

**Reply by jecampagne** [reply]
@jakevdp  the pb is rather at the level of 
```python
       #val1 = jax.lax.fori_loop(0,k,body_1, val1_init) # ok but certainly through while
       val1 = jec_fori_loop(0,k,body_1, val1_init)        # Nok scan failed due to Concrete
```

**Reply by jakevdp** [reply]
`k` is dynamic because it is derived from `m` and `mwrap`, which are dynamic due to being part of `carry`, which is always dynamic.

If you want `fori_loop` to lower to `scan` and be compatible with `jacrev`, you need the endpoints to be static. Anything in the `carry` or derived from it will be dynamic, so you need to move these things outside the carry in your function call. Fortunately it looks like the relevant quantities are constant across your nested loops, so this shouldn't pose a problem (aside from some refactoring).

**Reply by jecampagne** [reply]
Well @jakevdp  I understand the words but cannot get a solution to my case.  Let me elaborate a little. 
As a reminder
```python
k = jax.lax.min(m-i, mwrap-2)
```
So 

1.  `k` is a function of (`m`,`mwrap`) coming from the  `func_outer` function which `scan` the function `body_fn` (in the snippet I mention length=1 but of course this is not just one in real case) and so to pass  (`m`,`mwrap`) the only way is through the `carry` of the `scan`. Then  (`m`,`mwrap`)  is pass to the `func_inner` which uses another scan, etc.

2. but `k` is a function of `i` which is initialized in the `func_inner` as `i = mwrap-1`, and used in the `body_fn` of the `scan` in  `func_inner`. Notice that `i` is updated in the `step_once` (once in the snippet but many times in the real use case) and used in `jax.lax.cond(i<m,... which control the call to `step_once`

So already 1. in a problem. Do you see a simple work rounds?



More generally this schema cannot work but is it solvable?
```python
def finner(arg):
   
   def body_fn2(carry2,dummy):
        k = <from i,arg through carry2>
        jax.lax.fori_loop(0,k,... )                      # will use while_loop
        return carry2, None 
   
   i = <from arg>
   carry2_init = (i, arg)
   carry2, _ = jax.lax.scan(body_fn2, carry2_init,...)
   return ...

def fouter():

     def body_fn1(carry1,dummy):
          arg  = <extracted from carry>
         ... = finner(arg)
        carry1 = ...
        return carry1, None
                
      carry1_init=...
      carry1, _ = jax.lax.scan(body_fn1, carry1_init, None, length=n)
     
```

**Reply by jakevdp** [reply]
It’s hard for me to know how to adapt this pseudocode. I’ll reiterate the main principle: if you want loop endpoints to be static, they cannot be derived from dynamic quantities like the scan carry. Hope that makes sense

**Reply by jakevdp** [reply]
If you can provide a minimal working example (I.e. full code that executes), I might be able to give more pointed advice.

**Reply by jecampagne** [reply]
Hi @jakevdp , thanks fro your help. I understand that pseudo-code is not clear and I had already some trouble to get it... Well the originla code is not too long and is available in Colab [HERE](https://colab.research.google.com/drive/1C33PIy7Z5-XMrTeICKMx1lGWQn5YGaWr?usp=sharing) you can see that

1.  there is a peice of original code (_Code done with scan/while/fori loops_) that is working (ie does not crash) and provide some outputs when you execute `img_arr = test()` and see the img_arr array.
2. the code that I am trying to get running with scan-only (and cond) is the one that is under the section _Transform lax.fori_loop and lax.while in favour of lax.scan & lax.cond_ , you will see that it crashes for the reason you give me of course. So, if you have an idea to get it working, it would be very help full.

**Reply by jakevdp** [reply]
Thanks for that context. Looking at the full code, I don't think it's possible to write it in terms of static loops, because the size of the inner loop changes from iteration to iteration. Namely, `k = jax.lax.min(m-i, mwrap-2)` and `i` is updated each loop, and so the size of the inner loop is not static.

Your options would be either:

1. use forward-mode autodiff rather than reverse-mode autodiff with your dynamic loops
2. re-express your computation so that the size of the inner loop does not change from iteration to iteration

Hope that helps!

**Reply by jecampagne** [reply]
Ok thanks for your expertise. I will discuss with my colleagues a long your advise.
